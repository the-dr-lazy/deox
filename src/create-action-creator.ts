import { action, Action, AnyAction } from './action'

export type ActionCreator<T extends AnyAction | string> = T extends AnyAction
  ? {
      (...args: any[]): T
      type: T['type']
      toString(): T['type']
    }
  : (T extends string
      ? {
          (...args: any[]): Action<T>
          type: T
          toString(): T
        }
      : never)

/**
 * Flux standard action creator factory
 * @example
 * createAction('[Todo] truncate');
 * @example
 * createAction(
 *   '[Todo] fetch rejected',
 *   resolve => (error: Error) => resolve(error)
 * );
 * @example
 * createAction(
 *   '[Todo] fetch rejected',
 *   resolve => (error: Error, meta?: { status: number }) => resolve(error, meta)
 * )
 * @example
 * createAction(
 *   '[Todo] add',
 *   resolve => (name: string, completed = false) => resolve({ name, completed })
 * )
 * @example
 * createAction(
 *   '[Todo] add',
 *   resolve => (name: string, completed = false) => resolve({ name, completed }, 'Meta data of all todos')
 * )
 *
 */
export function createActionCreator<
  Type extends string,
  Callable extends <T>(...args: any[]) => Action<Type> = () => Action<Type>
>(
  type: Type,
  executor: (
    resolve: <Payload = undefined, Meta = undefined>(
      payload?: Payload,
      meta?: Meta
    ) => Action<Type, Payload, Meta>
  ) => Callable = resolve => (() => resolve()) as any
) {
  const callable = executor((payload, meta) => action(type, payload!, meta!))

  return Object.assign(callable, {
    type,
    toString() {
      return type
    },
  })
}
