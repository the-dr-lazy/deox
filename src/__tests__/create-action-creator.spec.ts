import { createActionCreator } from '../create-action-creator'

describe('createActionCreator', () => {
  it('should expose action type of action creator via .type', () => {
    const type = '[Todo] truncate'

    expect(createActionCreator(type).type).toBe(type)
  })

  it('should expose action type of action creator via .toString()', () => {
    const type = '[Todo] truncate'

    expect(createActionCreator(type).toString()).toBe(type)
  })

  it('should create action creator with type-only action without executor', () => {
    const actionCreator = createActionCreator('[Todo] truncate')

    expect(actionCreator()).toMatchSnapshot()
  })

  it('should create action creator with type-only action with empty executor', () => {
    const actionCreator = createActionCreator(
      '[Todo] truncate',
      resolve => () => resolve()
    )

    expect(actionCreator()).toMatchSnapshot()
  })

  it('should create action creator with error payload', () => {
    const actionCreator = createActionCreator(
      '[Todo] fetch rejected',
      resolve => (error: Error) => resolve(error)
    )

    expect(actionCreator(new Error())).toMatchSnapshot()
  })

  it('should create action creator with error payload and meta', () => {
    const actionCreator = createActionCreator(
      '[Todo] fetch rejected',
      resolve => (error: Error, meta: { status: number }) =>
        resolve(error, meta)
    )

    expect(actionCreator(new Error(), { status: 403 })).toMatchSnapshot()
  })

  it('should create action creator with non-error payload', () => {
    const actionCreator = createActionCreator(
      '[Todo] add',
      resolve => (name: string, completed = false) =>
        resolve({ name, completed })
    )

    expect(
      actionCreator('Prepare for practical parasitology course!')
    ).toMatchSnapshot()
  })

  it('should create action creator with non-error payload and meta', () => {
    const actionCreator = createActionCreator(
      '[Todo] add',
      resolve => (name: string, meta: string) => resolve({ name }, meta)
    )

    expect(
      actionCreator('Prepare for practical parasitology course!', 'Be patient')
    ).toMatchSnapshot()
  })
})
