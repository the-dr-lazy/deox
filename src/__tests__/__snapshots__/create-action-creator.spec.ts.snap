// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createActionCreator should create action creator with error payload 1`] = `
Object {
  "error": true,
  "payload": [Error],
  "type": "[Todo] fetch rejected",
}
`;

exports[`createActionCreator should create action creator with error payload and meta 1`] = `
Object {
  "error": true,
  "meta": Object {
    "status": 403,
  },
  "payload": [Error],
  "type": "[Todo] fetch rejected",
}
`;

exports[`createActionCreator should create action creator with non-error payload 1`] = `
Object {
  "payload": Object {
    "completed": false,
    "name": "Prepare for practical parasitology course!",
  },
  "type": "[Todo] add",
}
`;

exports[`createActionCreator should create action creator with non-error payload and meta 1`] = `
Object {
  "meta": "Be patient",
  "payload": Object {
    "name": "Prepare for practical parasitology course!",
  },
  "type": "[Todo] add",
}
`;

exports[`createActionCreator should create action creator with type-only action with empty executor 1`] = `
Object {
  "type": "[Todo] truncate",
}
`;

exports[`createActionCreator should create action creator with type-only action without executor 1`] = `
Object {
  "type": "[Todo] truncate",
}
`;
